[feat] 판매글 생성 API 구현

assokk opened the pull request:
## #️⃣ Issue Number<!--- ex) #이슈번호, #이슈번호 --> 

- Closes #76 

## 📝 요약(Summary)<!--- 변경 사항 및 관련 이슈에 대해 간단하게 작성해주세요. 어떻게보다 무엇을 왜 수정했는지 설명해주세요. -->
- **Endpoint**: `POST /api/sale-posts`
- **Request**: 제목, 내용, 가격, 카테고리, 이미지 URL 리스트
- **Response**: 생성된 판매글 정보
- @AuthenticationPrincipal로 사용자 식별, 추후 인증 방식 수정 예정
- S3 업로드는 추후 구현 예정 (현재는 URL로 받음)

## 🛠️ PR 유형

어떤 변경 사항이 있나요?

- [x] 새로운 기능 추가
- [ ] 버그 수정
- [ ] CSS 등 사용자 UI 디자인 변경
- [ ] 코드에 영향을 주지 않는 변경사항(오타 수정, 탭 사이즈 변경, 변수명 변경)
- [x] 코드 리팩토링
- [ ] 주석 추가 및 수정
- [ ] 문서 수정
- [ ] 테스트 추가, 테스트 리팩토링
- [ ] 빌드 부분 혹은 패키지 매니저 수정
- [ ] 파일 혹은 폴더명 수정
- [ ] 파일 혹은 폴더 삭제

## 📸스크린샷 (선택)## 💬 공유사항 to 리뷰어<!--- 리뷰어가 중점적으로 봐줬으면 좋겠는 부분이 있으면 적어주세요. -->
<!--- 논의해야할 부분이 있다면 적어주세요.-->
<!--- ex) 메서드 XXX의 이름을 더 잘 짓고 싶은데 혹시 좋은 명칭이 있을까요? -->

## ✅ PR Checklist

PR이 다음 요구 사항을 충족하는지 확인하세요.

- [ ] 커밋 메시지 컨벤션에 맞게 작성했습니다.
- [ ] 변경 사항에 대한 테스트를 했습니다.(버그 수정/기능에 대한 테스트).

assokk committed:
[feat] SalePost: 엔티티 및 예외 처리 구현

1. SalePost 엔티티 수정
  - User, Category, SalePostImage 연관관계 설정
  - 가격/이미지 검증 로직 추가
  - 이미지 연관관계 편의 메서드 추가
2. SalePostImage 엔티티 생성
  - SalePost 연관관계 설정
  - 이미지 URL 검증 로직 추가
3. 예외 처리
  - SalePostException, SalePostErrorCode 생성

assokk committed:
[feat] SalePost: 생성 API용 DTO 구현

- SalePostCreateRequest: 제목, 내용, 가격, 카테고리 검증
- SalePostCreateResponse: 생성된 판매글 정보 응답 (이미지 순서 정렬 포함)

assokk committed:
Merge remote-tracking branch 'origin' into feat/salePost

assokk committed:
[feat] SalePost: 생성 API용 Service 계층 구현

- SalePostCommandService: 판매글 생성 로직
  - User/Category 검증
  - 이미지 포함 판매글 생성

assokk committed:
[feat] SalePost: 생성 API Controller 구현

- SalePostCreateRequest에 imageUrls 필드 추가
- SalePostSuccessCode 생성
- SalePostController 구현

assokk committed:
[feat] SalePost: 조회 관련 코드 추가

- SalePostQueryService: findSalePostById 구현 (채팅 도메인 연동용)
- SalePostErrorCode: SALE_POST_NOT_FOUND 추가

assokk committed:
Merge remote-tracking branch 'origin' into feat/salePost-76

zerone1202 commented on src/main/java/org/example/ootoutfitoftoday/domain/salepost/exception/SalePostErrorCode.java:
```diff
@@ -1,4 +1,26 @@
+    INVALID_PRICE("INVALID_PRICE", HttpStatus.BAD_REQUEST, "가격은 0보다 커야 합니다."),
+    EMPTY_IMAGES("EMPTY_IMAGES", HttpStatus.BAD_REQUEST, "이미지는 최소 1개 이상 필요합니다."),
+    EMPTY_IMAGE_URL("EMPTY_IMAGE_URL", HttpStatus.BAD_REQUEST, "이미지 URL은 필수입니다."),
+    INVALID_DISPLAY_ORDER("INVALID_DISPLAY_ORDER", HttpStatus.BAD_REQUEST, "이미지 "),
```
이 부분은 수정이 필요해보이는데, 혹시 이미지로 해두신 이유가 있는걸까용?!

assokk replied:
리뷰 반영 완료

assokk replied:
ㅋㅋㅋㅋㅋㅋㅋㅋㅋ 뭐라고 쓸지 고민하다가 비운채로 올려버렸습니다

assokk commented on src/main/java/org/example/ootoutfitoftoday/domain/salepost/controller/SalePostController.java:
```diff
@@ -1,4 +1,38 @@
+            @AuthenticationPrincipal AuthUser authUser,
+            @Valid @RequestBody SalePostCreateRequest request
+    ) {
+        SalePostCreateResponse response = salePostCommandService.createSalePost(
+                authUser.getUserId(),
+                request,
+                request.getImageUrls()
+        );
```
추후 S3를 이용한 이미지 업로드 구현을 고려하여 역할을 분리해두었습니다. 수정이 필요해지면 그때 고려해보겠습니다!

assokk commented on src/main/java/org/example/ootoutfitoftoday/domain/salepost/dto/response/SalePostCreateResponse.java:
```diff
@@ -0,0 +1,43 @@
+
+    public static SalePostCreateResponse from(SalePost salePost) {
+
+        return SalePostCreateResponse.builder()
+                .salePostId(salePost.getId())
+                .title(salePost.getTitle())
+                .content(salePost.getContent())
+                .price(salePost.getPrice())
+                .status(salePost.getStatus().name())
+                .categoryId(salePost.getCategory().getId())
+                .imageUrls(salePost.getImages().stream()
+                        .sorted(Comparator.comparing(SalePostImage::getDisplayOrder))
+                        .map(SalePostImage::getImageUrl)
+                        .toList())
+                .createdAt(salePost.getCreatedAt())
+                .build();
```
리뷰 반영 완료

assokk commented on src/main/java/org/example/ootoutfitoftoday/domain/salepost/entity/SalePost.java:
```diff
@@ -32,45 +39,81 @@ public class SalePost extends BaseEntity {
+    @JoinColumn(name = "category_id", nullable = false)
+    private Category category;
+
+    @OneToMany(mappedBy = "salePost", cascade = CascadeType.ALL, orphanRemoval = true)
```
리뷰 반영 완료

assokk commented on src/main/java/org/example/ootoutfitoftoday/domain/salepost/entity/SalePost.java:
```diff
@@ -32,45 +39,81 @@ public class SalePost extends BaseEntity {
                 .status(SaleStatus.SELLING)
                 .build();
+
+        // 이미지 URL 리스트를 순서대로 SalePostImage 엔티티로 변환 (displayOrder: 1, 2, 3, ...)
+        for (int i = 0; i < imageUrls.size(); i++) {
+            SalePostImage image = SalePostImage.create(imageUrls.get(i), i + 1);
+            salePost.addImage(image);
+        }
```
싫어요현재 코드의 가독성이 더 좋고, 인덱스 기반 순서 부여 로직이 직관적으로 나타난다고 판단하여 반영하지 않겠습니다.
다만, 카운터 변수 방식이 더 자연스러워 보여 변경을 고려하고 있습니다.

zerone1202 reviewed (approved):
미지님, 수정까지 모두 확인했습니다! 수고많으셨습니다~

assokk committed:
[refactor] SalePost: 코드 리뷰 반영

- SalePost 엔티티에 @OrderBy 추가로 이미지 자동 정렬
- SalePostCreateResponse: 정렬 중복 로직 제거, userId 추가
- SalePostErrorCode 에러메세지 완성

zerone1202 reviewed (approved):
추가로 코드 리뷰 반영하신 부분 확인했습니다!
